File: src/settings_dialog.c
#include "settings_dialog.h"
#include <string.h>

struct _SettingsDialog {
    GtkWidget *dialog;
    GtkWidget *notebook;
    
    // Timer tab widgets
    GtkWidget *work_duration_spin;
    GtkWidget *short_break_spin;
    GtkWidget *long_break_spin;
    GtkWidget *sessions_spin;
    
    // Misc tab widgets
    GtkWidget *auto_start_check;
    GtkWidget *enable_sounds_check;
    GtkWidget *enable_idle_detection_check;
    GtkWidget *idle_timeout_spin;
    GtkWidget *idle_timeout_box;
    
    // Dialog buttons
    GtkWidget *restore_defaults_button;
    GtkWidget *cancel_button;
    GtkWidget *ok_button;
    
    SettingsDialogCallback callback;
    gpointer user_data;
};

static void on_restore_defaults_clicked(GtkButton *button, SettingsDialog *dialog);
static void on_cancel_clicked(GtkButton *button, SettingsDialog *dialog);
static void on_ok_clicked(GtkButton *button, SettingsDialog *dialog);
static void on_idle_detection_toggled(GtkToggleButton *button, SettingsDialog *dialog);

SettingsDialog* settings_dialog_new(GtkWindow *parent, const Settings *settings, AudioManager *audio) {
    (void)audio; // Not used in simplified version
    SettingsDialog *dialog = g_malloc0(sizeof(SettingsDialog));
    
    // Create dialog
    dialog->dialog = gtk_dialog_new();
    gtk_window_set_title(GTK_WINDOW(dialog->dialog), "Settings");
    gtk_window_set_transient_for(GTK_WINDOW(dialog->dialog), parent);
    gtk_window_set_modal(GTK_WINDOW(dialog->dialog), TRUE);
    gtk_window_set_default_size(GTK_WINDOW(dialog->dialog), 400, 300);
    gtk_window_set_resizable(GTK_WINDOW(dialog->dialog), FALSE);
    
    // Create main container
    GtkWidget *content_area = gtk_dialog_get_content_area(GTK_DIALOG(dialog->dialog));
    
    // Create notebook for tabs
    dialog->notebook = gtk_notebook_new();
    gtk_container_add(GTK_CONTAINER(content_area), dialog->notebook);
    
    // Create Timer tab
    GtkWidget *timer_tab = gtk_box_new(GTK_ORIENTATION_VERTICAL, 10);
    gtk_container_set_border_width(GTK_CONTAINER(timer_tab), 20);
    
    GtkWidget *timer_label = gtk_label_new("Timer");
    gtk_notebook_append_page(GTK_NOTEBOOK(dialog->notebook), timer_tab, timer_label);
    
    // Timer Durations section
    GtkWidget *durations_frame = gtk_frame_new("Timer Durations");
    gtk_box_pack_start(GTK_BOX(timer_tab), durations_frame, FALSE, FALSE, 0);
    
    GtkWidget *durations_grid = gtk_grid_new();
    gtk_grid_set_row_spacing(GTK_GRID(durations_grid), 10);
    gtk_grid_set_column_spacing(GTK_GRID(durations_grid), 10);
    gtk_container_set_border_width(GTK_CONTAINER(durations_grid), 15);
    gtk_container_add(GTK_CONTAINER(durations_frame), durations_grid);
    
    // Work Duration
    GtkWidget *work_label = gtk_label_new("Work Duration:");
    gtk_widget_set_halign(work_label, GTK_ALIGN_START);
    gtk_grid_attach(GTK_GRID(durations_grid), work_label, 0, 0, 1, 1);
    
    dialog->work_duration_spin = gtk_spin_button_new_with_range(1, 120, 1);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->work_duration_spin), settings->work_duration);
    gtk_grid_attach(GTK_GRID(durations_grid), dialog->work_duration_spin, 1, 0, 1, 1);
    
    GtkWidget *work_min_label = gtk_label_new("min");
    gtk_grid_attach(GTK_GRID(durations_grid), work_min_label, 2, 0, 1, 1);
    
    // Short Break
    GtkWidget *short_break_label = gtk_label_new("Short Break:");
    gtk_widget_set_halign(short_break_label, GTK_ALIGN_START);
    gtk_grid_attach(GTK_GRID(durations_grid), short_break_label, 0, 1, 1, 1);
    
    dialog->short_break_spin = gtk_spin_button_new_with_range(1, 60, 1);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->short_break_spin), settings->short_break_duration);
    gtk_grid_attach(GTK_GRID(durations_grid), dialog->short_break_spin, 1, 1, 1, 1);
    
    GtkWidget *short_min_label = gtk_label_new("min");
    gtk_grid_attach(GTK_GRID(durations_grid), short_min_label, 2, 1, 1, 1);
    
    // Long Break
    GtkWidget *long_break_label = gtk_label_new("Long Break:");
    gtk_widget_set_halign(long_break_label, GTK_ALIGN_START);
    gtk_grid_attach(GTK_GRID(durations_grid), long_break_label, 0, 2, 1, 1);
    
    dialog->long_break_spin = gtk_spin_button_new_with_range(5, 120, 1);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->long_break_spin), settings->long_break_duration);
    gtk_grid_attach(GTK_GRID(durations_grid), dialog->long_break_spin, 1, 2, 1, 1);
    
    GtkWidget *long_min_label = gtk_label_new("min");
    gtk_grid_attach(GTK_GRID(durations_grid), long_min_label, 2, 2, 1, 1);
    
    // Sessions until long break
    GtkWidget *sessions_label = gtk_label_new("Sessions until Long Break:");
    gtk_widget_set_halign(sessions_label, GTK_ALIGN_START);
    gtk_grid_attach(GTK_GRID(durations_grid), sessions_label, 0, 3, 1, 1);
    
    dialog->sessions_spin = gtk_spin_button_new_with_range(2, 10, 1);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->sessions_spin), settings->sessions_until_long_break);
    gtk_grid_attach(GTK_GRID(durations_grid), dialog->sessions_spin, 1, 3, 1, 1);
    
    // Create Misc tab
    GtkWidget *misc_tab = gtk_box_new(GTK_ORIENTATION_VERTICAL, 10);
    gtk_container_set_border_width(GTK_CONTAINER(misc_tab), 20);
    
    GtkWidget *misc_label = gtk_label_new("Misc");
    gtk_notebook_append_page(GTK_NOTEBOOK(dialog->notebook), misc_tab, misc_label);
    
    // Behavior Settings
    GtkWidget *behavior_frame = gtk_frame_new("Behavior Settings");
    gtk_box_pack_start(GTK_BOX(misc_tab), behavior_frame, FALSE, FALSE, 0);
    
    GtkWidget *behavior_box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 10);
    gtk_container_set_border_width(GTK_CONTAINER(behavior_box), 15);
    gtk_container_add(GTK_CONTAINER(behavior_frame), behavior_box);
    
    dialog->auto_start_check = gtk_check_button_new_with_label("Auto-start work after breaks end");
    gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(dialog->auto_start_check), settings->auto_start_work_after_break);
    gtk_widget_set_margin_bottom(dialog->auto_start_check, 8);
    gtk_box_pack_start(GTK_BOX(behavior_box), dialog->auto_start_check, FALSE, FALSE, 0);
    
    dialog->enable_sounds_check = gtk_check_button_new_with_label("Enable sound alerts");
    gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(dialog->enable_sounds_check), settings->enable_sounds);
    gtk_box_pack_start(GTK_BOX(behavior_box), dialog->enable_sounds_check, FALSE, FALSE, 0);
    
    // Idle detection section
    gtk_widget_set_margin_top(dialog->enable_sounds_check, 8);
    dialog->enable_idle_detection_check = gtk_check_button_new_with_label("Auto-pause when idle");
    gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(dialog->enable_idle_detection_check), settings->enable_idle_detection);
    gtk_widget_set_margin_top(dialog->enable_idle_detection_check, 8);
    gtk_box_pack_start(GTK_BOX(behavior_box), dialog->enable_idle_detection_check, FALSE, FALSE, 0);
    
    // Idle timeout configuration
    dialog->idle_timeout_box = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 10);
    gtk_widget_set_margin_left(dialog->idle_timeout_box, 24);
    gtk_widget_set_margin_top(dialog->idle_timeout_box, 5);
    gtk_box_pack_start(GTK_BOX(behavior_box), dialog->idle_timeout_box, FALSE, FALSE, 0);
    
    GtkWidget *idle_timeout_label = gtk_label_new("Idle timeout:");
    gtk_box_pack_start(GTK_BOX(dialog->idle_timeout_box), idle_timeout_label, FALSE, FALSE, 0);
    
    dialog->idle_timeout_spin = gtk_spin_button_new_with_range(1, 30, 1);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->idle_timeout_spin), settings->idle_timeout_minutes);
    gtk_box_pack_start(GTK_BOX(dialog->idle_timeout_box), dialog->idle_timeout_spin, FALSE, FALSE, 0);
    
    GtkWidget *idle_min_label = gtk_label_new("minutes");
    gtk_box_pack_start(GTK_BOX(dialog->idle_timeout_box), idle_min_label, FALSE, FALSE, 0);
    
    // Set sensitivity based on checkbox state
    gtk_widget_set_sensitive(dialog->idle_timeout_box, settings->enable_idle_detection);
    
    // Connect toggle signal to enable/disable timeout spinner
    g_signal_connect(dialog->enable_idle_detection_check, "toggled", 
                     G_CALLBACK(on_idle_detection_toggled), dialog);
    
    // Dialog buttons
    GtkWidget *action_area = gtk_dialog_get_action_area(GTK_DIALOG(dialog->dialog));
    
    dialog->restore_defaults_button = gtk_button_new_with_label("Restore Defaults");
    gtk_container_add(GTK_CONTAINER(action_area), dialog->restore_defaults_button);
    
    dialog->cancel_button = gtk_button_new_with_label("Cancel");
    gtk_container_add(GTK_CONTAINER(action_area), dialog->cancel_button);
    
    dialog->ok_button = gtk_button_new_with_label("OK");
    gtk_style_context_add_class(gtk_widget_get_style_context(dialog->ok_button), "suggested-action");
    gtk_container_add(GTK_CONTAINER(action_area), dialog->ok_button);
    
    // Connect signals
    g_signal_connect(dialog->restore_defaults_button, "clicked", G_CALLBACK(on_restore_defaults_clicked), dialog);
    g_signal_connect(dialog->cancel_button, "clicked", G_CALLBACK(on_cancel_clicked), dialog);
    g_signal_connect(dialog->ok_button, "clicked", G_CALLBACK(on_ok_clicked), dialog);
    
    return dialog;
}

void settings_dialog_free(SettingsDialog *dialog) {
    if (!dialog) return;
    
    gtk_widget_destroy(dialog->dialog);
    g_free(dialog);
}

void settings_dialog_set_callback(SettingsDialog *dialog, SettingsDialogCallback callback, gpointer user_data) {
    if (!dialog) return;
    
    dialog->callback = callback;
    dialog->user_data = user_data;
}

void settings_dialog_show(SettingsDialog *dialog) {
    if (!dialog) return;
    
    gtk_widget_show_all(dialog->dialog);
}

Settings* settings_dialog_get_settings(SettingsDialog *dialog) {
    if (!dialog) return NULL;
    
    Settings *settings = g_malloc0(sizeof(Settings));
    
    // Timer settings
    settings->work_duration = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(dialog->work_duration_spin));
    settings->short_break_duration = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(dialog->short_break_spin));
    settings->long_break_duration = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(dialog->long_break_spin));
    settings->sessions_until_long_break = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(dialog->sessions_spin));
    
    // Behavior settings
    settings->auto_start_work_after_break = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(dialog->auto_start_check));
    settings->enable_idle_detection = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(dialog->enable_idle_detection_check));
    settings->idle_timeout_minutes = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(dialog->idle_timeout_spin));
    
    // Audio settings (simplified)
    settings->enable_sounds = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(dialog->enable_sounds_check));
    settings->sound_volume = 0.7; // Fixed reasonable volume
    settings->sound_type = g_strdup("chimes");
    settings->work_start_sound = NULL;
    settings->break_start_sound = NULL;
    settings->session_complete_sound = NULL;
    settings->timer_finish_sound = NULL;
    
    return settings;
}

Settings* settings_new_default(void) {
    Settings *settings = g_malloc0(sizeof(Settings));
    
    settings->work_duration = 25;
    settings->short_break_duration = 5;
    settings->long_break_duration = 15;
    settings->sessions_until_long_break = 4;
    settings->auto_start_work_after_break = TRUE;
    settings->enable_idle_detection = FALSE;  // Off by default
    settings->idle_timeout_minutes = 2;        // 2 minutes default
    settings->enable_sounds = TRUE;
    settings->sound_volume = 0.7; // Fixed reasonable volume
    settings->sound_type = g_strdup("chimes");
    settings->work_start_sound = NULL;
    settings->break_start_sound = NULL;
    settings->session_complete_sound = NULL;
    settings->timer_finish_sound = NULL;
    
    return settings;
}

void settings_free(Settings *settings) {
    if (!settings) return;
    
    g_free(settings->sound_type);
    g_free(settings->work_start_sound);
    g_free(settings->break_start_sound);
    g_free(settings->session_complete_sound);
    g_free(settings->timer_finish_sound);
    g_free(settings);
}

Settings* settings_copy(const Settings *settings) {
    if (!settings) return NULL;
    
    Settings *copy = g_malloc0(sizeof(Settings));
    
    copy->work_duration = settings->work_duration;
    copy->short_break_duration = settings->short_break_duration;
    copy->long_break_duration = settings->long_break_duration;
    copy->sessions_until_long_break = settings->sessions_until_long_break;
    copy->auto_start_work_after_break = settings->auto_start_work_after_break;
    copy->enable_idle_detection = settings->enable_idle_detection;
    copy->idle_timeout_minutes = settings->idle_timeout_minutes;
    copy->enable_sounds = settings->enable_sounds;
    copy->sound_volume = settings->sound_volume;
    copy->sound_type = g_strdup(settings->sound_type);
    copy->work_start_sound = g_strdup(settings->work_start_sound);
    copy->break_start_sound = g_strdup(settings->break_start_sound);
    copy->session_complete_sound = g_strdup(settings->session_complete_sound);
    copy->timer_finish_sound = g_strdup(settings->timer_finish_sound);
    
    return copy;
}

// Callback implementations
static void on_restore_defaults_clicked(GtkButton *button, SettingsDialog *dialog) {
    (void)button; // Suppress unused parameter warning
    
    Settings *defaults = settings_new_default();
    
    // Reset all controls to defaults
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->work_duration_spin), defaults->work_duration);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->short_break_spin), defaults->short_break_duration);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->long_break_spin), defaults->long_break_duration);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->sessions_spin), defaults->sessions_until_long_break);
    gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(dialog->auto_start_check), defaults->auto_start_work_after_break);
    gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(dialog->enable_sounds_check), defaults->enable_sounds);
    gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(dialog->enable_idle_detection_check), defaults->enable_idle_detection);
    gtk_spin_button_set_value(GTK_SPIN_BUTTON(dialog->idle_timeout_spin), defaults->idle_timeout_minutes);
    
    settings_free(defaults);
}

static void on_cancel_clicked(GtkButton *button, SettingsDialog *dialog) {
    (void)button; // Suppress unused parameter warning
    
    if (dialog->callback) {
        dialog->callback("cancel", dialog->user_data);
    }
}

static void on_ok_clicked(GtkButton *button, SettingsDialog *dialog) {
    (void)button; // Suppress unused parameter warning
    
    if (dialog->callback) {
        dialog->callback("ok", dialog->user_data);
    }
}

static void on_idle_detection_toggled(GtkToggleButton *button, SettingsDialog *dialog) {
    gboolean active = gtk_toggle_button_get_active(button);
    gtk_widget_set_sensitive(dialog->idle_timeout_box, active);
}

File: src/gui/settings_dialog.zig
const std = @import("std");
const cw = @import("../c_workaround.zig");
const c = cw.c;
const Settings = @import("../config/settings.zig").Settings;

pub const SettingsDialog = struct {
    allocator: std.mem.Allocator,
    
    // GTK Widgets
    dialog: *c.GtkWidget,
    notebook: *c.GtkWidget,
    
    // Timer tab widgets
    work_duration_spin: *c.GtkWidget,
    short_break_spin: *c.GtkWidget,
    long_break_spin: *c.GtkWidget,
    sessions_spin: *c.GtkWidget,
    
    // Misc tab widgets
    auto_start_check: *c.GtkWidget,
    enable_sounds_check: *c.GtkWidget,
    enable_idle_detection_check: *c.GtkWidget,
    idle_timeout_spin: *c.GtkWidget,
    idle_timeout_box: *c.GtkWidget,
    
    // Dialog buttons
    restore_defaults_button: *c.GtkWidget,
    cancel_button: *c.GtkWidget,
    ok_button: *c.GtkWidget,
    
    // Callback
    callback: ?*const fn (action: []const u8, user_data: ?*anyopaque) void,
    user_data: ?*anyopaque,
    
    // Current settings
    settings: *Settings,
    
    const Self = @This();
    
    pub fn init(allocator: std.mem.Allocator, parent: ?*c.GtkWindow, settings: *Settings) !*Self {
        const dialog_self = try allocator.create(Self);
        errdefer allocator.destroy(dialog_self);
        
        // Create dialog
        const dialog = c.gtk_dialog_new() orelse return error.DialogCreationFailed;
        c.gtk_window_set_title(@ptrCast(dialog), "Settings");
        if (parent) |p| {
            c.gtk_window_set_transient_for(@ptrCast(dialog), p);
        }
        c.gtk_window_set_modal(@ptrCast(dialog), c.TRUE);
        c.gtk_window_set_default_size(@ptrCast(dialog), 400, 300);
        c.gtk_window_set_resizable(@ptrCast(dialog), c.FALSE);
        
        // Get content area
        const content_area = c.gtk_dialog_get_content_area(@ptrCast(dialog));
        
        // Create notebook for tabs
        const notebook = c.gtk_notebook_new() orelse return error.WidgetCreationFailed;
        c.gtk_container_add(@ptrCast(content_area), notebook);
        
        // Create Timer tab
        const timer_tab = c.gtk_box_new(@intFromEnum(c.GtkOrientation.GTK_ORIENTATION_VERTICAL), 10) orelse return error.WidgetCreationFailed;
        c.gtk_container_set_border_width(@ptrCast(timer_tab), 20);
        
        const timer_label = c.gtk_label_new("Timer") orelse return error.WidgetCreationFailed;
        _ = c.gtk_notebook_append_page(@ptrCast(notebook), timer_tab, timer_label);
        
        // Timer Durations section
        const durations_frame = c.gtk_frame_new("Timer Durations") orelse return error.WidgetCreationFailed;
        c.gtk_box_pack_start(@ptrCast(timer_tab), durations_frame, c.FALSE, c.FALSE, 0);
        
        const durations_grid = c.gtk_grid_new() orelse return error.WidgetCreationFailed;
        c.gtk_grid_set_row_spacing(@ptrCast(durations_grid), 10);
        c.gtk_grid_set_column_spacing(@ptrCast(durations_grid), 10);
        c.gtk_container_set_border_width(@ptrCast(durations_grid), 15);
        c.gtk_container_add(@ptrCast(durations_frame), durations_grid);
        
        // Work Duration
        const work_label = c.gtk_label_new("Work Duration:") orelse return error.WidgetCreationFailed;
        c.gtk_widget_set_halign(work_label, @intFromEnum(c.GtkAlign.GTK_ALIGN_START));
        c.gtk_grid_attach(@ptrCast(durations_grid), work_label, 0, 0, 1, 1);
        
        const work_duration_spin = c.gtk_spin_button_new_with_range(1, 120, 1) orelse return error.WidgetCreationFailed;
        c.gtk_spin_button_set_value(@ptrCast(work_duration_spin), @floatFromInt(settings.work_duration));
        c.gtk_grid_attach(@ptrCast(durations_grid), work_duration_spin, 1, 0, 1, 1);
        
        const work_min_label = c.gtk_label_new("min") orelse return error.WidgetCreationFailed;
        c.gtk_grid_attach(@ptrCast(durations_grid), work_min_label, 2, 0, 1, 1);
        
        // Short Break
        const short_break_label = c.gtk_label_new("Short Break:") orelse return error.WidgetCreationFailed;
        c.gtk_widget_set_halign(short_break_label, @intFromEnum(c.GtkAlign.GTK_ALIGN_START));
        c.gtk_grid_attach(@ptrCast(durations_grid), short_break_label, 0, 1, 1, 1);
        
        const short_break_spin = c.gtk_spin_button_new_with_range(1, 60, 1) orelse return error.WidgetCreationFailed;
        c.gtk_spin_button_set_value(@ptrCast(short_break_spin), @floatFromInt(settings.short_break_duration));
        c.gtk_grid_attach(@ptrCast(durations_grid), short_break_spin, 1, 1, 1, 1);
        
        const short_min_label = c.gtk_label_new("min") orelse return error.WidgetCreationFailed;
        c.gtk_grid_attach(@ptrCast(durations_grid), short_min_label, 2, 1, 1, 1);
        
        // Long Break
        const long_break_label = c.gtk_label_new("Long Break:") orelse return error.WidgetCreationFailed;
        c.gtk_widget_set_halign(long_break_label, @intFromEnum(c.GtkAlign.GTK_ALIGN_START));
        c.gtk_grid_attach(@ptrCast(durations_grid), long_break_label, 0, 2, 1, 1);
        
        const long_break_spin = c.gtk_spin_button_new_with_range(5, 120, 1) orelse return error.WidgetCreationFailed;
        c.gtk_spin_button_set_value(@ptrCast(long_break_spin), @floatFromInt(settings.long_break_duration));
        c.gtk_grid_attach(@ptrCast(durations_grid), long_break_spin, 1, 2, 1, 1);
        
        const long_min_label = c.gtk_label_new("min") orelse return error.WidgetCreationFailed;
        c.gtk_grid_attach(@ptrCast(durations_grid), long_min_label, 2, 2, 1, 1);
        
        // Sessions until long break
        const sessions_label = c.gtk_label_new("Sessions until Long Break:") orelse return error.WidgetCreationFailed;
        c.gtk_widget_set_halign(sessions_label, @intFromEnum(c.GtkAlign.GTK_ALIGN_START));
        c.gtk_grid_attach(@ptrCast(durations_grid), sessions_label, 0, 3, 1, 1);
        
        const sessions_spin = c.gtk_spin_button_new_with_range(2, 10, 1) orelse return error.WidgetCreationFailed;
        c.gtk_spin_button_set_value(@ptrCast(sessions_spin), @floatFromInt(settings.sessions_until_long_break));
        c.gtk_grid_attach(@ptrCast(durations_grid), sessions_spin, 1, 3, 1, 1);
        
        // Create Misc tab
        const misc_tab = c.gtk_box_new(@intFromEnum(c.GtkOrientation.GTK_ORIENTATION_VERTICAL), 10) orelse return error.WidgetCreationFailed;
        c.gtk_container_set_border_width(@ptrCast(misc_tab), 20);
        
        const misc_label = c.gtk_label_new("Misc") orelse return error.WidgetCreationFailed;
        _ = c.gtk_notebook_append_page(@ptrCast(notebook), misc_tab, misc_label);
        
        // Behavior Settings
        const behavior_frame = c.gtk_frame_new("Behavior Settings") orelse return error.WidgetCreationFailed;
        c.gtk_box_pack_start(@ptrCast(misc_tab), behavior_frame, c.FALSE, c.FALSE, 0);
        
        const behavior_box = c.gtk_box_new(@intFromEnum(c.GtkOrientation.GTK_ORIENTATION_VERTICAL), 10) orelse return error.WidgetCreationFailed;
        c.gtk_container_set_border_width(@ptrCast(behavior_box), 15);
        c.gtk_container_add(@ptrCast(behavior_frame), behavior_box);
        
        const auto_start_check = c.gtk_check_button_new_with_label("Auto-start work after breaks end") orelse return error.WidgetCreationFailed;
        c.gtk_toggle_button_set_active(@ptrCast(auto_start_check), if (settings.auto_start_work_after_break) c.TRUE else c.FALSE);
        c.gtk_widget_set_margin_bottom(auto_start_check, 8);
        c.gtk_box_pack_start(@ptrCast(behavior_box), auto_start_check, c.FALSE, c.FALSE, 0);
        
        const enable_sounds_check = c.gtk_check_button_new_with_label("Enable sound alerts") orelse return error.WidgetCreationFailed;
        c.gtk_toggle_button_set_active(@ptrCast(enable_sounds_check), if (settings.enable_sounds) c.TRUE else c.FALSE);
        c.gtk_box_pack_start(@ptrCast(behavior_box), enable_sounds_check, c.FALSE, c.FALSE, 0);
        
        // Idle detection section
        c.gtk_widget_set_margin_top(enable_sounds_check, 8);
        const enable_idle_detection_check = c.gtk_check_button_new_with_label("Auto-pause when idle") orelse return error.WidgetCreationFailed;
        c.gtk_toggle_button_set_active(@ptrCast(enable_idle_detection_check), if (settings.enable_idle_detection) c.TRUE else c.FALSE);
        c.gtk_widget_set_margin_top(enable_idle_detection_check, 8);
        c.gtk_box_pack_start(@ptrCast(behavior_box), enable_idle_detection_check, c.FALSE, c.FALSE, 0);
        
        // Idle timeout configuration
        const idle_timeout_box = c.gtk_box_new(@intFromEnum(c.GtkOrientation.GTK_ORIENTATION_HORIZONTAL), 10) orelse return error.WidgetCreationFailed;
        c.gtk_widget_set_margin_left(idle_timeout_box, 24);
        c.gtk_widget_set_margin_top(idle_timeout_box, 5);
        c.gtk_box_pack_start(@ptrCast(behavior_box), idle_timeout_box, c.FALSE, c.FALSE, 0);
        
        const idle_timeout_label = c.gtk_label_new("Idle timeout:") orelse return error.WidgetCreationFailed;
        c.gtk_box_pack_start(@ptrCast(idle_timeout_box), idle_timeout_label, c.FALSE, c.FALSE, 0);
        
        const idle_timeout_spin = c.gtk_spin_button_new_with_range(1, 30, 1) orelse return error.WidgetCreationFailed;
        c.gtk_spin_button_set_value(@ptrCast(idle_timeout_spin), @floatFromInt(settings.idle_timeout_minutes));
        c.gtk_box_pack_start(@ptrCast(idle_timeout_box), idle_timeout_spin, c.FALSE, c.FALSE, 0);
        
        const idle_min_label = c.gtk_label_new("minutes") orelse return error.WidgetCreationFailed;
        c.gtk_box_pack_start(@ptrCast(idle_timeout_box), idle_min_label, c.FALSE, c.FALSE, 0);
        
        // Set sensitivity based on checkbox state
        c.gtk_widget_set_sensitive(idle_timeout_box, if (settings.enable_idle_detection) c.TRUE else c.FALSE);
        
        // Dialog buttons
        const action_area = c.gtk_dialog_get_action_area(@ptrCast(dialog));
        
        const restore_defaults_button = c.gtk_button_new_with_label("Restore Defaults") orelse return error.WidgetCreationFailed;
        c.gtk_container_add(@ptrCast(action_area), restore_defaults_button);
        
        const cancel_button = c.gtk_button_new_with_label("Cancel") orelse return error.WidgetCreationFailed;
        c.gtk_container_add(@ptrCast(action_area), cancel_button);
        
        const ok_button = c.gtk_button_new_with_label("OK") orelse return error.WidgetCreationFailed;
        c.gtk_style_context_add_class(c.gtk_widget_get_style_context(ok_button), "suggested-action");
        c.gtk_container_add(@ptrCast(action_area), ok_button);
        
        dialog_self.* = .{
            .allocator = allocator,
            .dialog = dialog,
            .notebook = notebook,
            .work_duration_spin = work_duration_spin,
            .short_break_spin = short_break_spin,
            .long_break_spin = long_break_spin,
            .sessions_spin = sessions_spin,
            .auto_start_check = auto_start_check,
            .enable_sounds_check = enable_sounds_check,
            .enable_idle_detection_check = enable_idle_detection_check,
            .idle_timeout_spin = idle_timeout_spin,
            .idle_timeout_box = idle_timeout_box,
            .restore_defaults_button = restore_defaults_button,
            .cancel_button = cancel_button,
            .ok_button = ok_button,
            .callback = null,
            .user_data = null,
            .settings = settings,
        };
        
        // Connect signals
        _ = c.g_signal_connect_data(restore_defaults_button, "clicked", @ptrCast(&onRestoreDefaultsClicked), dialog_self, null, 0);
        _ = c.g_signal_connect_data(cancel_button, "clicked", @ptrCast(&onCancelClicked), dialog_self, null, 0);
        _ = c.g_signal_connect_data(ok_button, "clicked", @ptrCast(&onOkClicked), dialog_self, null, 0);
        _ = c.g_signal_connect_data(enable_idle_detection_check, "toggled", @ptrCast(&onIdleDetectionToggled), dialog_self, null, 0);
        
        return dialog_self;
    }
    
    pub fn deinit(self: *Self) void {
        c.gtk_widget_destroy(self.dialog);
        self.allocator.destroy(self);
    }
    
    pub fn setCallback(self: *Self, callback: *const fn ([]const u8, ?*anyopaque) void, user_data: ?*anyopaque) void {
        self.callback = callback;
        self.user_data = user_data;
    }
    
    pub fn show(self: *Self) void {
        c.gtk_widget_show_all(self.dialog);
    }
    
    pub fn getSettings(self: *Self) Settings {
        var new_settings = Settings.init(self.allocator);
        
        // Get values from spinners
        new_settings.work_duration = @intFromFloat(c.gtk_spin_button_get_value(@ptrCast(self.work_duration_spin)));
        new_settings.short_break_duration = @intFromFloat(c.gtk_spin_button_get_value(@ptrCast(self.short_break_spin)));
        new_settings.long_break_duration = @intFromFloat(c.gtk_spin_button_get_value(@ptrCast(self.long_break_spin)));
        new_settings.sessions_until_long_break = @intFromFloat(c.gtk_spin_button_get_value(@ptrCast(self.sessions_spin)));
        
        // Get checkbox states
        new_settings.auto_start_work_after_break = c.gtk_toggle_button_get_active(@ptrCast(self.auto_start_check)) == c.TRUE;
        new_settings.enable_sounds = c.gtk_toggle_button_get_active(@ptrCast(self.enable_sounds_check)) == c.TRUE;
        new_settings.enable_idle_detection = c.gtk_toggle_button_get_active(@ptrCast(self.enable_idle_detection_check)) == c.TRUE;
        new_settings.idle_timeout_minutes = @intFromFloat(c.gtk_spin_button_get_value(@ptrCast(self.idle_timeout_spin)));
        
        return new_settings;
    }
    
    fn updateFromSettings(self: *Self, settings: *const Settings) void {
        // Update spinners
        c.gtk_spin_button_set_value(@ptrCast(self.work_duration_spin), @floatFromInt(settings.work_duration));
        c.gtk_spin_button_set_value(@ptrCast(self.short_break_spin), @floatFromInt(settings.short_break_duration));
        c.gtk_spin_button_set_value(@ptrCast(self.long_break_spin), @floatFromInt(settings.long_break_duration));
        c.gtk_spin_button_set_value(@ptrCast(self.sessions_spin), @floatFromInt(settings.sessions_until_long_break));
        
        // Update checkboxes
        c.gtk_toggle_button_set_active(@ptrCast(self.auto_start_check), if (settings.auto_start_work_after_break) c.TRUE else c.FALSE);
        c.gtk_toggle_button_set_active(@ptrCast(self.enable_sounds_check), if (settings.enable_sounds) c.TRUE else c.FALSE);
        c.gtk_toggle_button_set_active(@ptrCast(self.enable_idle_detection_check), if (settings.enable_idle_detection) c.TRUE else c.FALSE);
        c.gtk_spin_button_set_value(@ptrCast(self.idle_timeout_spin), @floatFromInt(settings.idle_timeout_minutes));
    }
    
    // Event handlers
    fn onRestoreDefaultsClicked(button: *c.GtkButton, user_data: c.gpointer) callconv(.C) void {
        _ = button;
        const self: *Self = @ptrCast(@alignCast(user_data));
        
        // Create default settings and update UI
        var defaults = Settings.init(self.allocator);
        self.updateFromSettings(&defaults);
        
        if (self.callback) |cb| {
            cb("restore_defaults", self.user_data);
        }
    }
    
    fn onCancelClicked(button: *c.GtkButton, user_data: c.gpointer) callconv(.C) void {
        _ = button;
        const self: *Self = @ptrCast(@alignCast(user_data));
        
        c.gtk_widget_hide(self.dialog);
        
        if (self.callback) |cb| {
            cb("cancel", self.user_data);
        }
    }
    
    fn onOkClicked(button: *c.GtkButton, user_data: c.gpointer) callconv(.C) void {
        _ = button;
        const self: *Self = @ptrCast(@alignCast(user_data));
        
        c.gtk_widget_hide(self.dialog);
        
        if (self.callback) |cb| {
            cb("ok", self.user_data);
        }
    }
    
    fn onIdleDetectionToggled(button: *c.GtkToggleButton, user_data: c.gpointer) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(user_data));
        const active = c.gtk_toggle_button_get_active(button) == c.TRUE;
        c.gtk_widget_set_sensitive(self.idle_timeout_box, if (active) c.TRUE else c.FALSE);
    }
};